File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/CommandFrameBase.vb
'*******************************************************************************************************
'  CommandFrameBase.vb - Command frame base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Text
Imports Tva.Phasors.Common

''' <summary>This class represents the protocol independent common implementation of a command frame that can be sent or received from a PMU.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class CommandFrameBase

    Inherits ChannelFrameBase(Of ICommandCell)
    Implements ICommandFrame

    Private m_command As DeviceCommand

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize command frame
        m_command = info.GetValue("command", GetType(DeviceCommand))

    End Sub

    Protected Sub New(ByVal cells As CommandCellCollection, ByVal command As DeviceCommand)

        MyBase.New(cells)
        m_command = command

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    ' and automatically pass in parsing state
    Protected Sub New(ByVal state As ICommandFrameParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(state, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal commandFrame As ICommandFrame)
    Protected Sub New(ByVal commandFrame As ICommandFrame)

        MyClass.New(commandFrame.Cells, commandFrame.Command)

    End Sub

    Protected Overrides ReadOnly Property FundamentalFrameType() As FundamentalFrameType
        Get
            Return Phasors.FundamentalFrameType.CommandFrame
        End Get
    End Property

    Public Overridable Shadows ReadOnly Property Cells() As CommandCellCollection Implements ICommandFrame.Cells
        Get
            Return MyBase.Cells
        End Get
    End Property

    Public Overridable Property Command() As DeviceCommand Implements ICommandFrame.Command
        Get
            Return m_command
        End Get
        Set(ByVal value As DeviceCommand)
            m_command = value
        End Set
    End Property

    Public Overridable Property ExtendedData() As Byte() Implements ICommandFrame.ExtendedData
        Get
            Return Cells.BinaryImage
        End Get
        Set(ByVal value As Byte())
            Cells.Clear()
            MyBase.ParseBodyImage(New CommandFrameParsingState(Cells, 0, value.Length), value, 0)
        End Set
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            Return MyBase.BodyLength + 2
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = CreateArray(Of Byte)(BodyLength)
            Dim index As Int32 = 2

            EndianOrder.BigEndian.CopyBytes(m_command, buffer, 0)
            CopyImage(MyBase.BodyImage, buffer, index, MyBase.BodyLength)

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Int32)

        m_command = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
        MyBase.ParseBodyImage(state, binaryImage, startIndex + 2)

    End Sub

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize command frame
        info.AddValue("command", m_command, GetType(DeviceCommand))

    End Sub

    Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
        Get
            With MyBase.Attributes
                .Add("Device Command", Command & ": " & [Enum].GetName(GetType(DeviceCommand), Command))

                If Cells.Count > 0 Then
                    .Add("Extended Data", ByteEncoding.Hexadecimal.GetString(Cells.BinaryImage, 0, Cells.Count))
                Else
                    .Add("Extended Data", "<null>")
                End If
            End With

            Return MyBase.Attributes
        End Get
    End Property

End Class

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/CommandFrameBase.cs
//*******************************************************************************************************
//  CommandFrameBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  01/14/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Text;
using System.Runtime.Serialization;
using System.Collections.Generic;

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the protocol independent common implementation of any <see cref="ICommandFrame"/> that can be sent or received.
    /// </summary>
    [Serializable()]
    public abstract class CommandFrameBase : ChannelFrameBase<ICommandCell>, ICommandFrame
    {
        #region [ Members ]

        // Fields
        private DeviceCommand m_command;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="CommandFrameBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected CommandFrameBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize command frame
            m_command = (DeviceCommand)info.GetValue("command", typeof(DeviceCommand));
        }

        /// <summary>
        /// Creates a new <see cref="CommandFrameBase"/> from the specified parameters.
        /// </summary>
        protected CommandFrameBase(CommandCellCollection cells, DeviceCommand command)
            : base(0, cells, 0)
        {
            m_command = command;
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="FundamentalFrameType"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public override FundamentalFrameType FrameType
        {
            get
            {
                return FundamentalFrameType.CommandFrame;
            }
        }

        /// <summary>
        /// Gets reference to the <see cref="CommandCellCollection"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public virtual new CommandCellCollection Cells
        {
            get
            {
                return base.Cells as CommandCellCollection;
            }
        }

        /// <summary>
        /// Gets or sets <see cref="DeviceCommand"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public virtual DeviceCommand Command
        {
            get
            {
                return m_command;
            }
            set
            {
                m_command = value;
            }
        }

        /// <summary>
        /// Gets or sets extended binary image data for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public virtual byte[] ExtendedData
        {
            get
            {
                return Cells.BinaryImage;
            }
            set
            {
                Cells.Clear();
                State = new CommandFrameParsingState(0, value.Length);
                ParseBodyImage(value, 0, value.Length);
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="HeaderImage"/>.
        /// </summary>
        protected override int HeaderImage
        {
            get
            {
                return 2;
            }
        }

        /// <summary>
        /// Gets the binary header image of this <see cref="CommandFrameBase"/>.
        /// </summary>
        protected override byte[] HeaderImage
        {
            get
            {
                byte[] buffer = new byte[2];

                EndianOrder.BigEndian.CopyBytes((short)m_command, buffer, 0);

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="CommandFrameBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Device Command", (int)Command + ": " + Command);

                if (Cells.Count > 0)
                    baseAttributes.Add("Extended Data", ByteEncoding.Hexadecimal.GetString(Cells.BinaryImage, 0, Cells.Count));
                else
                    baseAttributes.Add("Extended Data", "<null>");

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Parses the binary header image.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        protected override int ParseHeaderImage(byte[] binaryImage, int startIndex, int length)
        {
            m_command = (DeviceCommand)EndianOrder.BigEndian.ToInt16(binaryImage, startIndex);
            return 2;
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize command frame
            info.AddValue("command", m_command, typeof(DeviceCommand));
        }

        #endregion
    }
}